sim/xcelium/__init__.py

```diff
@@ -0,0 +1,394 @@
+#!/usr/bin/env python3
+
+# HAMMER-VLSI PLUGIN, XCELIUM 
+# Notes: This plugin sets up xrun to execute in a three-step xrun invocation.
+#        This bridges multi-tool direct-invocation and the xrun single-invocation.
+
+
+import os
+import re
+import shutil
+import json
+import datetime
+import io
+import sys
+from typing import Dict, List, Optional, Callable, Tuple
+from multiprocessing import Process
+
+import hammer_utils
+import hammer_tech
+from hammer_tech import HammerTechnologyUtils
+from hammer_vlsi import FlowLevel, TimeValue
+from hammer_vlsi import HammerSimTool, HammerToolStep, HammerLSFSubmitCommand, HammerLSFSettings
+from hammer_logging import HammerVLSILogging
+
+sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)),"../../common"))
+from tool import CadenceTool
+
+class xcelium(HammerSimTool, CadenceTool):
+
+  @property
+  def xcelium_ext(self) -> List[str]:
+    verilog_ext  = [".v", ".V", ".VS", ".vp", ".VP"]
+    sverilog_ext = [".sv",".SV",".svp",".SVP",".svi",".svh",".vlib",".VLIB"]
+    c_cxx_ext    = [".c",".cc",".cpp"]
+    gz_ext       = [ext + ".gz" for ext in verilog_ext + sverilog_ext]
+    z_ext        = [ext + ".z" for ext  in verilog_ext + sverilog_ext]
+    return (verilog_ext + sverilog_ext + c_cxx_ext + gz_ext + z_ext)
+
+  @property
+  def steps(self) -> List[HammerToolStep]:
+    return self.make_steps_from_methods([self.compile_xrun,
+                                         self.elaborate_xrun,
+                                         self.sim_xrun])
+    
+  @property
+  def tool_config_prefix(self) -> str:
+    return "sim.xcelium"
+  
+  @property
+  def sim_input_prefix(self) -> str:
+    return "sim.inputs"
+  
+  @property
+  def sim_waveform_prefix(self) -> str:
+    return "sim.inputs.waveform"
+  
+  @property
+  def xcelium_bin(self) -> str:
+    return self.get_setting("sim.xcelium.xcelium_bin")
+
+  @property
+  def sim_tcl_file(self) -> str: 
+    return os.path.join(self.run_dir, "xrun_sim.tcl")
+    
+  @property
+  def sdf_cmd_file(self) -> str:
+    return os.path.join(self.run_dir, "design.sdf_cmd")
+  
+  def post_synth_sdc(self) -> Optional[str]:
+    pass
+
+  def write_gl_files(self) -> bool:
+    return True
+  
+  def get_verilog_models(self) -> List[str]:
+    verilog_sim_files = self.technology.read_libs([
+        hammer_tech.filters.verilog_sim_filter], 
+        hammer_tech.HammerTechnologyUtils.to_plain_item)
+    return verilog_sim_files
+        
+  def fill_outputs(self) -> bool:
+    self.output_waveforms = []
```

Comment:
Since you're controlling waveforms with input keys, I think this output waveforms list can be auto-populated. Ditto SAIF.

Reply:
If specified, fields are now autopopulated.